File: R/math.R
# Math required for pulse-specific, weighted relative amplitude
# estimation.

library(stats)


#' Returns a penalization factor for some z-score, such that
#'
#' @param x Subject ID
quad_penalty <- function(z, alpha = 1) {
    return(alpha / (z^2 + 1))
}

gaussian_penalty <- function(z, alpha = 1) {
    return(exp(-z^2 / alpha))
}
gaussian_penalty(-2, alpha = 10)
quad_penalty(3, alpha = 4)

robust_zscores <- function(v) {
    mad <- mad(v)
    rz <- 0.6745 * (v - median(v)) / mad

    return(rz)
}

File: jl/math.jl
function p(d::Number, t::Vector, f=Φ)::Number
    """Computes weighted, pulse-specific relative
    amplitude given a paired-pulse d and a series
    of test pulses t. Weights are computed via some 
    penalty f."""
    w = broadcast(f, rzs(t))
    p = (d ./ t) .* w
    sum(p) / sum(w)
end

function p(d::Vector, t::Vector, f=Φ)::Vector
    h(x) = p(x, t, f)
    broadcast(h, d)
end

function p(d::SubArray, t::Vector, f=Φ)::Vector
    h(x) = p(x, t, f)
    broadcast(h, d)
end


function rawp(d::Vector, t::Vector)::Vector
    d ./ mean(t)
end

function rawp(d::SubArray, t::Vector)::Vector
    d ./ mean(t)
end

function ℒ(θ::Vector, d::Vector, f)
    w = broadcast(f, d)
    sum(θ .* w) / sum(w)
end

function Φ(x, α=1)
    α / (x^2 + α)
end

function Ψ(x, α=1)
    exp(-(x^2 / α))
end

function ∅(x)
    x
end
function rzs(x::Vector)::Vector
    mad = median(abs.(x .- median(x)))
    0.675 * (x .- median(x)) / mad

end
